nameko_sqlalchemy/database_session.py

```diff
@@ -42,3 +42,17 @@ def worker_teardown(self, worker_ctx):
 
 # backwards compat
 Session = DatabaseSession
+
+
+def run_query(session, query, attempts=2):
```

Comment:
You can extract the `session` from the given `query` object to avoid passing it in. See http://docs.sqlalchemy.org/en/latest/orm/session_basics.html#how-can-i-get-the-session-for-a-certain-object

I also wonder if we can give this a nicer API by implementing it as a context manager?

Something like:

``` python
with transaction_retry(attempts=X):
    return self.session.query(...)
```

Reply:
I don't think that the `session` can be extracted because the `query` argument is a simple python function and not a sqlalchemy `query` object.
It would be nice to have this implemented as a context manager but in that case it would need to yield more than once and as far as I know it is not possible.

Reply:
The `query` argument, based on your test cases and docs, most of the time would be a bound method executing a query (`all`, `count`, ...) on an instance of `Query`. You can maybe default to it?

```
if not session:
    if isinstance(query.__self__, Query):
        session = query.__self__.session
    else:
        raise ValueError(...
```
It would work in cases you wrap methods executing a query:
```
run_query(self.session.query(SomeModel).all, attempts=2)
```
It wouldn't be that explicit to try to infer the session, but you know, "too explicit is not that good as explicit" ;)

Reply:
As your `run_query` wraps a function executing a query (or more queries) and your goal is to retry the whole transaction scope contained in that one function wrapped, then it seems a decorator would be a nice API for it including a nice name like the one @mattbennett suggested above:

```
@transaction_retry(attempts=X)
def my_transaction(self):
    self.session.add(...
    ...
```